src/backend/catalog/storage_pending_deletes_redo.c

```diff
@@ -0,0 +1,319 @@
+#include "postgres.h"
+
+#include "access/clog.h"
+#include "access/transam.h"
+#include "catalog/storage_pending_deletes.h"
+#include "miscadmin.h"
+#include "storage/md.h"
+#include "utils/elog.h"
+#include "utils/guc.h"
+#include "utils/hsearch.h"
+
+/*
+ * HTAB entry for pending deletes for the given xid.
+ */
+typedef struct PendingDeleteHtabNode
+{
+	TransactionId xid;
+	List	   *relnode_list;	/* list of RelFileNodePendingDelete */
+}	PendingDeleteHtabNode;
+
+/*
+ * Hash table for pending deletes relfilenodes for a given xid.
+ */
+static HTAB *pendingDeletesRedo = NULL;
+
+/*
+ * This function inserts XLOG_PENDING_DELETE record into WAL. It is called
+ * during checkpoint creation.
+ */
+void
+PdlXLogInsert()
+{
+	Size		size = 0;
+
+	if (IsBootstrapProcessingMode() || !gp_track_pending_delete)
+		return;
+
+	PendingRelXactDeleteArray *arr = PdlXLogShmemDump(&size);
+
+	if (size > 0 && arr != NULL)
+	{
+		XLogRecPtr	rec;
+		XLogRecData rdata;
+
+		rdata.buffer = InvalidBuffer;
+		rdata.data = (char *) arr;
+		rdata.len = size;
+		rdata.next = NULL;
+		rdata.buffer_std = false;
+
+		rec = XLogInsert(RM_XLOG_ID, XLOG_PENDING_DELETE, &rdata);
+
+		XLogFlush(rec);
+
+		elog(DEBUG1, "Pending delete XLog record inserted");
+
+		pfree(arr);
+	}
+}
+
+/*
+ * This function adds pending delete node to a pendingDeletesRedo hash-table
+ * during WAL redo processing. It is called when:
+ *	1. XLOG_SMGR_CREATE xlog message is replayed by smrg_redo;
+ *	2. XLOG_PENDING_DELETE xlog message is replayed by PdlRedoXLogRecord.
+ */
+void
+PdlRedoAdd(PendingRelXactDelete * pd)
+{
+	Assert(pd);
+
+	if (IsBootstrapProcessingMode() ||
+		(pd->xid == InvalidTransactionId) ||
+		!gp_track_pending_delete)
+		return;
+
+	if (NULL == pendingDeletesRedo)
+	{
+		HASHCTL		info =
+		{
+			.keysize = sizeof(TransactionId),
+			.entrysize = sizeof(PendingDeleteHtabNode)
+		};
+
+		pendingDeletesRedo = hash_create("pendingDeletesRedo hash",
+										 32,
+										 &info,
+										 HASH_ELEM);
+	}
+
+	bool		found = false;
+
+	PendingDeleteHtabNode *entry =
+	(PendingDeleteHtabNode *) hash_search(pendingDeletesRedo, &pd->xid, HASH_ENTER, &found);
+
+	if (!found)
+	{
+		entry->xid = pd->xid;
+		entry->relnode_list = NIL;
+	}
+
+	RelFileNodePendingDelete *data = (RelFileNodePendingDelete *) palloc(sizeof(*data));
+
+	memcpy(data, &pd->relnode, sizeof(*data));
```

Comment:
when we use memcpy, compiler doesn't check arguments types comportability. When we use `=`, compiler can generate more optimal code (do not call a fuction)
```suggestion
	*data = pd->relnode;
```